src/shared/ui/picker.ts

```diff
@@ -145,3 +148,213 @@ export function verifySinglePickerOutput<T extends vscode.QuickPickItem>(choices
+// If we move this to a cache, we should remove the awsCallLogic and instead listen/send events to/from the cache.
+// (this should also be retooled to not use the request/response objects directly)
+// TODO: Add manual pagination?
+/**
+ * Grants a `vscode.QuickPick` iteration capabilities:
+ * * Pause/unpause loading
+ * * Refresh capabilities
+ * * No item/error nodes
+ *
+ * External control is still done via `picker.promptUser` as if it were a normal QuickPick.
+ */
+export class IteratingQuickPickController<TResponse> {
+    private isDone: boolean = false
+    private isPaused: boolean = false
+    private iterator: AsyncIterator<vscode.QuickPickItem[]>
+    private activeIterationTime: Date
+
+    // Default constructs are public static so they can be validated aganist by other functions.
+    public static readonly REFRESH_BUTTON: vscode.QuickInputButton = {
+        iconPath: new vscode.ThemeIcon('refresh'),
+        tooltip: localize('AWS.generic.refresh', 'Refresh'),
+    }
+    public static readonly NO_ITEMS_ITEM: vscode.QuickPickItem = {
+        label: localize('AWS.picker.dynamic.noItemsFound.label', '[No items found]'),
+        detail: localize('AWS.picker.dynamic.noItemsFound.detail', 'Click here to go back'),
+        alwaysShow: true,
+    }
+    public static readonly ERROR_ITEM: vscode.QuickPickItem = {
+        label: localize('AWS.picker.dynamic.errorNode.label', 'There was an error retrieving more items.'),
+        alwaysShow: true,
+    }
+
+    /**
+     * @param quickPick A `vscode.QuickPick` to grant iterating capabilities
+     * @param populator An IteratingQuickPickPopulator which can call an iterator and return QuickPickItems.
+     */
+    public constructor(
+        private readonly quickPick: vscode.QuickPick<vscode.QuickPickItem>,
+        private readonly populator: IteratingQuickPickPopulator<TResponse>
+    ) {
+        // append buttons specific to iterating quickPick
+        this.quickPick.buttons = [...this.quickPick.buttons, IteratingQuickPickController.REFRESH_BUTTON]
+
+        this.iterator = this.populator.getPickIterator()
+
+        this.activeIterationTime = new Date()
+    }
+
+    /**
+     * Pauses any existing item loading.
+     * Call upon selecting an item in order to pause additional background loading.
+     */
+    public pauseRequests(): void {
+        this.isPaused = true
+    }
+
+    /**
+     * Starts item loading. Restarts from existing state if paused.
+     * Useful for manual pagination, losing quick pick focus (e.g. via external link), throttling, etc.
+     */
+    public startRequests(): void {
+        this.loadItems(false)
+    }
+
+    /**
+     * Resets quick pick's state. Useful for manual refreshes with the same query parameters
+     */
+    public reset(): void {
+        this.loadItems(true)
+    }
+
+    private async loadItems(reset: boolean): Promise<void> {
```

Comment:
Usually considered best practice not to use raw booleans as parameters when the meaning cannot be inferred from the function name. Here, it's not obvious that `loadItems(false)` has any effect on the rest.

Additionally, (resetting) and (loading) are two separate concerns that would be better split out into seperate functions.

Perhaps move the reset logic into the public `reset` method and call `loadItems()` afterwards.

Reply:
I can look but I did want to make this function specifically conscious of the reset state due to array handling (specifically, [line 251](https://github.com/aws/aws-toolkit-vscode/pull/1111/files#diff-93fe9b673d53886017037f1c38992b92R251)).

Reply:
If `reset` is called, you should be able to immediately set the quick pick items to an empty array before entering the `loadItems` method which seems to accomplish what you're after.

Reply:
The reason why I want to set it as late as possible is due the fact that other loadItems calls could still be in-flight. Since we're only modifying the quickPick's items directly, there's a chance that any actively-running loadItems calls from past refreshes could pollute the most recent version of the array. This is what the timestamp check does away with.